client/my-sites/plans-features-main/hooks/use-check-plan-availability-for-purchase.ts

```diff
@@ -0,0 +1,43 @@
+import { Plans } from '@automattic/data-stores';
+import { useSelector } from 'react-redux';
+import isPlanAvailableForPurchase from 'calypso/state/sites/plans/selectors/is-plan-available-for-purchase';
+import { getSelectedSiteId } from 'calypso/state/ui/selectors';
+import type { PlanSlug } from '@automattic/calypso-products';
+
+interface Props {
+	planSlugs: PlanSlug[];
+}
+
+type PlanAvailabilityForPurchase = {
+	[ planSlug in PlanSlug ]: boolean;
+};
+
+/**
+ * Returns a dictionary of plan slugs and whether or not they are available for purchase.
+ * Note that:
+ * - If there is not current-plan or selected-site defined, then plan is marked as purchasable.
+ * This would be the equivalent of an Onboarding/Signup context, where no other conditions currently apply.
+ * - If a selected site exists, then we check if the plan is available for purchase on that site.
+ */
+const useCheckPlanAvailabilityForPurchase = ( {
+	planSlugs,
+}: Props ): PlanAvailabilityForPurchase => {
+	const selectedSiteId = useSelector( getSelectedSiteId );
+	const currentPlan = Plans.useCurrentPlan( { siteId: selectedSiteId } );
+
+	return useSelector( ( state ) =>
+		Object.fromEntries(
+			planSlugs.map( ( planSlug ) => {
+				// 1. No selected site or current plan: plan is in a context-free state and is available for purchase
+				if ( ! selectedSiteId || ! currentPlan ) {
+					return [ planSlug, true ];
+				}
```

Comment:
@jeyip concerning your comment in the other PR:

> Checking for both siteID and currentPlan is curious to me ( although I understand that it's what was originally defined in usePricingMetaForGridPlans ). I'm wondering when that would occur 🤔

It's a good question. I can't think of anything other than a "site without a plan" making for a special case where plans become purchasable (as in Signup), or otherwise use of `currentPlan` being completely redundant. 🤷‍♂️ 

It would be good to get to the bottom of it really. I get almost always dizzy thinking about these conditions. We have a habit of building granular conditions to serve things as they come, while seeing the bigger picture (the whole) is not easy or possible. I keep hitting that impression personally quite often as I work on things. Maybe with these refactors we've been doing things will be easier to see and follow.

Reply:
> Maybe with these refactors we've been doing things will be easier to see and follow.

💯 

> It would be good to get to the bottom of it really.

YES! I'll do a bit of timeboxed investigating tomorrow whether or not this PR is merged by then. 👍

Reply:
> YES! I'll do a bit of timeboxed investigating tomorrow whether or not this PR is merged by then. 👍

Amazing! Let me know if you come up with something, or I'd be happy to prioritize review if you send PR instead (thinking of merging this soon since our testing approves)

Reply:
Starting to look now 🤞 Let's see if we can dig up some learnings here

Reply:
My investigations aren't definitive, but they do point to the idea that there's redundancy in checking `currentPlan`

The only real use case I could find that might potentially fit the bill of "site without a plan" were domain only "sites" P6Ya1B-gw-p2. I confirmed that there is an underlying wordpress.com domain for the domain only site. It also appears to exist without an attached plan.

As far as I could tell, however, there wasn't a situation where the user would ever see the plans grid _outside_ of signup or onboarding ( which means that checking `selectedSiteId` would be enough )

Reply:
Thanks @jeyip! Nice digging. 

So I hear that a "site without a plan" is a valid state, except that it currently exists in a scenario where "site" and "plan" cannot co-exist e.g. a plan doesn't apply to a "domain-only" site. However, if this becomes a valid state in Signup (e.g. a site is created prior to rendering the plans-grid), then "availability for purchase" would fall into the check below (assuming `selectedSiteID` becomes `true`), while we may instead want to have this be `true` (if there is also no plan attached/purchased on the site when created).

It sounds complex i.e. if a site is indeed created prior to rendering the plans-grid and there is a plan associated with it, then the "state-based check" below should capture that scenario too. I think? 😵

Reply:
> So I hear that a "site without a plan" is a valid state, except that it currently exists in a scenario where "site" and "plan" cannot co-exist e.g. a plan doesn't apply to a "domain-only" site.

Yeah this is exactly what I've seen so far 👍

> However, if this becomes a valid state in Signup (e.g. a site is created prior to rendering the plans-grid), then "availability for purchase" would fall into the check below (assuming selectedSiteID becomes true)

Oh this is an interesting observation. Come to think of it, I believe the domain upsell flow falls into this category ( user creates a site and revisits a task in Launchpad, which redirects them to domain upsell onboarding )

> It sounds complex i.e. if a site is indeed created prior to rendering the plans-grid and there is a plan associated with it, then the "state-based check" below should capture that scenario too (not sure if it does though). I think? 😵

I'm getting dizzy 😆 🤯

Let's try it out with the domain upsell. I think you're right. I'll test it out when I have some time tomorrow

Reply:
@chriskmnds Just double checked the domain upsell flow ( by completing the free flow and using launchpad )

<img width="1720" alt="Screenshot 2023-12-19 at 3 29 51 PM" src="https://github.com/Automattic/wp-calypso/assets/5414230/b0e223a6-1bce-4c33-824f-7ff7eb587281">

To summarize, a site is created with a free plan before they ever see a plans grid. They are then prompted to choose a domain, at which point they're redirected to the domain upsell flow. If a custom domain is chosen, they are shown the plans grid.

I've run into a hiccup. Everything appears to work as expected in production, but not locally 😆 Unsure what that's all about

### Local Dev
<img width="1707" alt="Screenshot 2023-12-19 at 3 49 02 PM" src="https://github.com/Automattic/wp-calypso/assets/5414230/c2b4059b-7847-4fe2-a62f-8c210d0836f9">

### Production
<img width="1724" alt="Screenshot 2023-12-19 at 3 50 51 PM" src="https://github.com/Automattic/wp-calypso/assets/5414230/c7356df4-8d5f-4bdb-9774-6914e6cbac64">

>  if a site is indeed created prior to rendering the plans-grid and there is a plan associated with it, then the "state-based check" below should capture that scenario too (not sure if it does though). I think?

Going to leave my investigation here for now, but to comment in this, it's what I would've figured too. I'm not certain either.